### PR TITLE
feat(debounce): Support AbortSignal to debounce for improved cancellation

### PR DESCRIPTION
**Pull request by Hanna922**
Hi👋🏻 I have reviewed issue #37 and thought it was a good feature, so I have added it.

> The [debounce](https://es-toolkit.slash.page/reference/function/debounce.html) function already supports cancellation through its return value, but its usability would improve if it also supported cancellation via [AbortSignal](https://developer.mozilla.org/en-US/docs/Web/API/AbortSignal), which is [implemented in all major web browsers and JavaScript runtimes](https://developer.mozilla.org/en-US/docs/Web/API/AbortSignal#browser_compatibility).

If this change is acceptable, I'll add docs and extend this functionality to the `delay` function.

### Before

```ts
function debounce<F extends (...args: any[]) => void>(func: F, debounceMs: number): F & { cancel: () => void }
```

### After

```ts
function debounce<F extends (...args: any[]) => void>(
  func: F,
  debounceMs: number,
  abortSignal?: AbortSignal
): F & { cancel: () => void }
```

### CODE REVIEW

**Comment by raon0211** [comment]
Hello @Hanna922, Thanks for your contribution!

File: src/function/debounce.ts [hunk 1]
```diff
@@ -17,25 +18,57 @@
...
+ * }, 1000, signal);
+ *
+ * // Will cancel the debounced function call
+ * controller.abort();
  */
-export function debounce<F extends (...args: any[]) => void>(func: F, debounceMs: number): F & { cancel: () => void } {
+export function debounce<F extends (...args: any[]) => void>(
+  func: F,
+  debounceMs: number,
+  abortSignal?: AbortSignal
```
Comment (raon0211) [inline src/function/debounce.ts:35]: To receive signals, what about we follow the [fetch API's option](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch#aborting_a_fetch)?

```tsx
interface DebounceOptions {
  signal?: AbortSignal;
}

export function debounce<F extends (...args: any[]) => void>(
  func: F,
  debounceMs: number,
  { signal }: DebounceOptions = {}
) { /* ... */ }
```

File: src/function/debounce.ts [hunk 2]
```diff
@@ -17,25 +18,57 @@
...
   debounced.cancel = function () {
-    if (timeoutId != null) {
+    if (timeoutId !== null) {
       clearTimeout(timeoutId);
+      timeoutId = null;
+    }
+
+    if (abortSignal) {
+      abortSignal.removeEventListener('abort', onAbort);
     }
```
Comment (raon0211) [inline src/function/debounce.ts:66]:
```suggestion
    abortSignal?.removeEventListener('abort', onAbort);
```

What about using optional chaining here?

File: src/function/debounce.ts [hunk 3]
```diff
@@ -17,25 +18,57 @@
...
   };
 
+  if (abortSignal) {
+    abortSignal.addEventListener('abort', onAbort);
+  }
```
Comment (raon0211) [inline src/function/debounce.ts:71]:
```suggestion
  abortSignal?.addEventListener('abort', onAbort);
```

Here too

Comment (Hanna922) [reply src/function/debounce.ts]: It was reflected in '8cf4772'

File: src/function/debounce.ts (hunk 1 above)
Comment (Hanna922) [reply src/function/debounce.ts]: It was reflected in [b6ed097](https://github.com/toss/es-toolkit/pull/45/commits/b6ed097b8c2d62b77b42e736ccb5a57d08f63c0a)

**Comment by Hanna922** [comment]
@raon0211 I have applied the comments and added documentation! Thank you 🙇🏻‍♀️

Would it be good to apply these changes to throttle as well?

**Review by raon0211 — APPROVED** [review]
Awesome work!

**Pull request merged**